### ISSUE TITLE
`kornia.core` dropped all 32 of its `__all__` names in 0.8.3 with no deprecation shim and no release note (824 binding losses; #3986 is one of them)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4024]
At v0.8.2, `kornia/core/__init__.py` re-exported `._backend` and declared **32 names in `__all__`**:

```
Device Dtype Module ModuleList Parameter Sequential Tensor arange as_tensor complex
concatenate cos deg2rad diag einsum eye linspace map_coordinates normalize ones
ones_like pad rad2deg rand sin softmax stack tan tensor where zeros zeros_like
```

In 0.8.3, `kornia/core/_backend.py` was deleted and all 32 went with it — no `__getattr__` shim, no `@deprecated` wrapper, no release-note entry.

## Repro

```python
import kornia, kornia.core
print(kornia.__version__)
print([n for n in ("Tensor", "Module", "pad", "tensor", "zeros", "stack", "Device")
       if not hasattr(kornia.core, n)])
```

## Observed

```
0.9.0rc1
['Tensor', 'Module', 'pad', 'tensor', 'zeros', 'stack', 'Device']
```

All 20 spot-checked names are absent. `kornia/core/__init__.py` has no module-level `__getattr__`.

## Scale

Because ~500 modules did `from kornia.core import ...`, the removal took **824 module-scope bindings** with it across the tree. Most-lost: `Tensor` ×247, `Module` ×113, `tensor` ×59, `stack` ×43, `zeros` ×35, `concatenate` ×32, `as_tensor` ×27, `where` ×25, `pad` ×19.

#3986 is one of those 19 `pad` bindings — the one that happened to have a downstream consumer loud enough to file a bug (ComfyUI-LTXVideo, broken for three months across five ignored PRs).

`kornia.utils` went through the same removal in the same release and **did it correctly**: a module-level `__getattr__` emitting `DeprecationWarning` and naming `kornia.onnx.download`. All 13 spot-checked `kornia.utils` names still resolve. So the mechanism existed and was simply not applied here.

## Was this a policy violation?

**No, and that is worth stating plainly.** `docs/source/get-started/stability.rst` says a symbol is public when it appears in the rendered documentation, and that "undocumented helpers are private regardless of importability". `docs/source/core.rst` documents exactly two things: `TensorWrapper` and `kornia.core.utils.batched_forward`. None of the 32 was ever documented, so none was public API and promise 1 (deprecation before removal) did not attach.

What *was* missed is promise 3 — release notes carry the ledger. Nothing in `CHANGELOG.md` records this removal.

The uncomfortable part is that `kornia.core` is listed in the **stable core** tier on that same page, and its `__all__` named all 32. A user who reads `__all__` as the export declaration — which is what `__all__` is for — had every reason to treat them as public. The policy and the code disagreed, and the code is what people import.

## Proposal: do not resurrect these; fix the ledger and the ambiguity instead

Recommendation, for maintainer ratification:

1. **Leave all 32 removed.** Reasons: the removal was policy-compliant; two releases have now shipped without them (0.8.3, 0.9.0rc1) so the affected installed base has already pinned or migrated; and every one of the 32 is a plain torch name with a mechanical, lossless replacement (`Tensor` → `torch.Tensor`, `pad` → `torch.nn.functional.pad`, `concatenate` → `torch.cat`). There is no kornia-specific behavior to preserve. Re-adding them would restore the alias layer whose removal was the point, and then require breaking the same names a second time.
2. **Add the retroactive release-note entry** that promise 3 requires, with the mapping table.
3. **Close the `__all__`-vs-docs gap in the policy** so this cannot recur — see the companion change; `__all__` must not advertise names the policy considers private.

### The counter-evidence, stated honestly

A GitHub code search over public Python (sample of 60 non-fork hits) shows these names are genuinely imported downstream:

| name | repos observed | name | repos observed |
|---|---|---|---|
| `Tensor` | 39 | `zeros_like` | 2 |
| `stack` | 16 | `tensor` | 2 |
| `where` | 13 | `as_tensor` | 2 |
| `concatenate` | 13 | `Parameter` | 2 |
| `zeros` | 12 | `Dtype` | 2 |
| `ones_like` | 11 | `Device` | 2 |
| `Module` | 10 | `rand` | 1 |
| `eye` | 3 | `normalize` | 1 |

Named repos include `facebookresearch/vggsfm`, `farm-ng/farm-ng-amiga`, `leggedrobotics/self_supervised_segmentation`, `zju3dv/BoxDreamer`, `snuvclab/pegasus`, `snuvclab/perse`, `Luo-Yihang/3DEnhancer`, `Peter-Kocsis/IntrinsicImageFusion`, `liming-ai/ControlNet_Plus_Plus`. This is a sample, not a census — GitHub indexes a subset, so absence from the table is not evidence a name is unused.

**If maintainers prefer to soften the landing**, the minimal alternative is a `__getattr__` shim in `kornia/core/__init__.py` covering the 32 names, emitting `DeprecationWarning` that names the torch replacement, scheduled for removal one minor release out. It is roughly 15 lines and one dict. I do not recommend it — it re-opens a door that has been shut for two releases — but it is the option, and the decision belongs to maintainers, not to the audit.

## Classification

**API surface / process.** Not a behavior defect; the question is what the project owes an undocumented-but-declared export.

## Disposition

n/a — single mechanism, no duplicate pair.

## Environment

kornia 0.9.0rc1 @ `0685b1f5` (main), torch 2.12.0, python 3.11, cpu. Found by the export-surface audit opened from #3986. Related: #4023 (broken `__all__` entries), #3898 (the guard that would have caught this had `kornia.core` been baselined before the removal rather than after).

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Closing: **we are not resurrecting these.**

The missing piece of context is what `kornia.core` was for. It was an artifact of an attempt to make kornia usable across more than one backend — PyTorch and TensorFlow. `Tensor`, `Module`, `pad`, `zeros`, `stack` and the rest were indirection points so that the library could, in principle, be retargeted. That effort was abandoned, and once it was, the indirection had no purpose: every one of those 32 names was a pass-through to torch and nothing else.

So the 0.8.3 removal was not an oversight that happened to be policy-compliant. It was the cleanup that the abandoned backend work had left outstanding. Re-adding the names would restore an abstraction layer for a capability the project no longer pursues, and would then need removing a second time.

That settles the counter-evidence in the issue body too. The repos importing `from kornia.core import Tensor` are importing an alias for `torch.Tensor` — the migration is `from torch import Tensor`, it is mechanical, and it loses nothing.

What stands from this issue and is being carried forward:

- the release-note entry that promise 3 requires and never got written, with the mapping table;
- the point that `__all__` and the documentation page were free to disagree — `kornia.core.__all__` declared 32 symbols `docs/source/core.rst` never documented — which is what made the removal look like a broken promise to anyone reading `__all__` as the export declaration. Closing that gap is queued as policy work.

#4023 (three names still in `__all__` that do not resolve) is a live bug and stays open.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
